src/vision/shiny_future/alan_cancels_the_future.md

```diff
@@ -0,0 +1,201 @@
+
+```rust
+async fn do_the_thing() {
+    if check_database {
+        let conn = DbConnection::connect("sqlite::memory:").await?;
+        ...
+        conn.spawn_drop();
+    } // <-- time is being spent here, dropping `conn`
+}
+```
+
+The respone latency is looking much better now! "Still, that was a bit subtle," Grace thinks. "I wonder if there are other cases like that in my code."
+
+### Barbara's tale: Smooth interop with the rest of Rust
+
+#### Implementing AsyncDrop for the containers
+
+Barbara is working on extending the `Vec` type in the standard library to support async drop. The `Vec` type has a custom drop impl that first drops all of the items in the vector before freeing the vector's internal buffer. 
+
+```rust
+impl<T> AsyncDrop for BVec<T> {
+    async fn drop(&mut self) {
+
+    }
+}
+```
+
+#### Dropping in a sync context
+
+Barbara is working on a separate project in which she has a custom main. One of the first things her program does is to create a future `prepare_the_thing`; before launching it, though, the program sometimes errors out under some conditions.
+
+```rust
+fn main() {
+    let the_future = prepare_the_thing();
+
+    if some_condition {
+        eprintln!("Error: foo");
+        return;
+    }
+
+    custom_runtime::launch(the_future);
+}
+
+fn prepare_the_thing() -> impl Future<Output = ()> {
+    ...
+    async move { ... }
+}
+```
+
+As it happens, `prepare_the_thing` returns a future that has ownership of various items that implement `AsyncDrop`. While reading over the code, Barbara wonders, "What happens when this future is dropped synchronously?" Reading the docs for `AsyncDrop`, though, she learns that -- by default, at least -- the synchronous `Drop` invokes the asynchronous drop, using the simple `block_on` mechanism included in the standard library. Testing it, she finds that everything works as expected.
+
+She does wonder, "What if I didn't want `block_on` to block, what could I do then?" The book explains that she can implement the `Drop` trait explicitly and make it to do whatever she wants, including (for example) invoking `spawn_drop` to run in the background.
+
+### TODO
+
+* Work in some part of the story that addresses [the destructor state problem](https://boats.gitlab.io/blog/post/poll-drop/) from boat's blog post. How do we manage generic code that doesn't know whether the future produced by `Drop` is send, and how do we walk about that? Similarly trait objects?
+
+## 🤔 Frequently Asked Questions
+
+### What status quo stories are you retelling?
+
+*Link to status quo stories if they exist. If not, that's ok, we'll help find them.*
+
+### What are the key attributes of this shiny future?
+
+* There is an `AsyncDrop` trait that you can implement instead of `Drop`
+    * If you do so, `Drop` will execute your async drop with... `std::block_on`? Something.
```

Comment:
This implies some level of integration or compatibility between `std::block_on` and the async runtime. Specifically, the reactor needs to run for any kind of I/O to work.

I'm not sure where or whether to point out notes like this in the template, though.

Reply:
Yeah. The interaction between sync drop and async drop is pretty tricky. I'd like to see a few stories exploring different angles on it.